localstack-core/localstack/services/apigateway/next_gen/execute_api/handlers/api_key_validation.py

```diff
@@ -0,0 +1,115 @@
+import logging
+from typing import Optional
+
+from localstack.aws.api.apigateway import ApiKey, ApiKeySourceType, RestApi
+from localstack.http import Response
+
+from ..api import RestApiGatewayHandler, RestApiGatewayHandlerChain
+from ..context import InvocationRequest, RestApiInvocationContext
+from ..gateway_response import InvalidAPIKeyError
+from ..moto_helpers import get_api_key, get_usage_plan_keys, get_usage_plans
+from ..variables import ContextVarsIdentity
+
+LOG = logging.getLogger(__name__)
+
+
+class ApiKeyValidationHandler(RestApiGatewayHandler):
+    """
+    Handles Api key validation.
+    If an api key is required, we will validate that a usage plan associated with that stage
+    has a usage plan key with the corresponding value.
+    """
+
+    # TODO We currently do not support rate limiting or quota limit. As such we are not raising any related Exception
+
+    def __call__(
+        self,
+        chain: RestApiGatewayHandlerChain,
+        context: RestApiInvocationContext,
+        response: Response,
+    ):
+        method = context.resource_method
+        request = context.invocation_request
+        rest_api = context.deployment.rest_api.rest_api
+
+        # If api key is not required by the method, we can exit the handler
+        if not method.get("apiKeyRequired"):
+            return
+
+        # If the Identity context was not created yet, instantiate it and attach it to the context variables
+        if not (identity := context.context_variables.get("identity", ContextVarsIdentity())):
+            context.context_variables["identity"] = identity
+
+        # Look for the api key value in the request. If it is not found, raise an exception
+        if not (api_key_value := self.get_request_api_key(rest_api, request, identity)):
+            LOG.debug("API Key is empty")
+            raise InvalidAPIKeyError("Forbidden")
+
+        # Get the validated key, if no key is found, raise an exception
+        if not (validated_key := self.validate_api_key(api_key_value, context)):
+            LOG.debug("Provided API Key is not valid")
+            raise InvalidAPIKeyError("Forbidden")
+
+        # Update the context's identity with the key value and id
+        if not identity.get("apiKey"):
+            LOG.debug("Updating $context.identity.apiKey='%s'", validated_key["value"])
+            identity["apiKey"] = validated_key["value"]
+
+        LOG.debug("Updating $context.identity.apiKeyId='%s'", validated_key["id"])
+        identity["apiKeyId"] = validated_key["id"]
+
+    def validate_api_key(
```

Comment:
I believe once we migrate `UsagePlan` and `ApiKey` to our store, seeing this usage, we should keep a map of API Key values (+ maybe stage?) directly in order to quickly validate them? We could maybe then simplify this to not have to iterate over all usage plans and usage plan keys.

Reply:
Yes, I believe we can keep the item in the store in a way that will be more favourable to this specific access pattern. I don't think we will be able to avoid all looping, without creating some data duplication but we should be able to reduce the amount of work being done here.